overlord/restart/restart_context.go

```diff
@@ -0,0 +1,193 @@
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/release"
+)
+
+type RestartWaiter struct {
+	TaskID string
+	Status state.Status
+}
+
+// XXX: How should we handle bootinfo and multiple requesters
+// when passing to restart package. Right now we just pass the last one.
+type RestartContext struct {
+	Waiters        []*RestartWaiter
+	RestoreWaiters bool
+	SnapName       string
+	RestartType    RestartType
+	RebootInfo     *boot.RebootInfo
+	Requested      bool
+}
+
+func (rt *RestartContext) needsReboot() bool {
+	return len(rt.Waiters) > 0 && !rt.Requested
+}
+
+func (rt *RestartContext) setParameters(snapName string, restartType RestartType, rebootInfo *boot.RebootInfo) {
+	rt.Requested = false
+	rt.SnapName = snapName
+	rt.RestartType = restartType
+	rt.RebootInfo = rebootInfo
+}
+
+func (rt *RestartContext) reboot(st *state.State) {
+	rt.Requested = true
+	rt.Waiters = nil
+	rt.RestoreWaiters = false
+
+	if release.OnClassic {
+		// Notify the system that a reboot is required.
+		if err := notifyRebootRequiredClassic(rt.SnapName); err != nil {
+			logger.Noticef("cannot notify about pending reboot: %v", err)
+		}
+		logger.Noticef("Postponing restart until a manual system restart allows to continue")
+		return
+	}
+	Request(st, rt.RestartType, rt.RebootInfo)
+}
+
+func (rt *RestartContext) taskFromID(chg *state.Change, id string) *state.Task {
+	for _, t := range chg.Tasks() {
+		if t.ID() == id {
+			return t
+		}
+	}
+	return nil
+}
+
+func (rt *RestartContext) restoreWaiters(chg *state.Change) {
+	for _, w := range rt.Waiters {
+		if t := rt.taskFromID(chg, w.TaskID); t != nil {
+			t.SetStatus(w.Status)
+		}
+	}
+	rt.Waiters = nil
+	rt.RestoreWaiters = false
+}
+
+func (rt *RestartContext) doUndoReboot(t *state.Task, status state.Status) error {
+	// Immediately, on the first undo, we clear all the waiters that previously
+	// has requested a reboot, that are now in WaitStatus. We put those back into
+	// their original requested state.
+	if rt.RestoreWaiters {
+		rt.restoreWaiters(t.Change())
+	}
+	return rt.doReboot(t, status)
+}
+
+func (rt *RestartContext) taskInSlice(t *state.Task, ts []*state.Task) bool {
+	for _, tt := range ts {
+		if tt == t {
+			return true
+		}
+	}
+	return false
+}
+
+// markTaskForRestart sets certain properties on a task to mark it for restart.
+func (rt *RestartContext) markTaskForRestart(task *state.Task, snapName string, status state.Status) {
+	rm := restartManager(task.State(), "internal error: cannot request a restart before RestartManager initialization")
+
+	// store current boot id to be able to check later if we have rebooted or not
+	task.Set("wait-for-system-restart-from-boot-id", rm.bootID)
+	task.SetToWait(status)
+	setWaitForSystemRestart(task.Change())
+
+	rt.Waiters = append(rt.Waiters, &RestartWaiter{
+		TaskID: task.ID(),
+		Status: status,
+	})
+}
+
+func (rt *RestartContext) doReboot(t *state.Task, status state.Status) error {
```

Comment:
doReboot is a strange name for this it's very confusing wrt reboot()

this also need an explanation of the arguments and intention